From 046ccb896bd365b9f8eab3734722f20fa1e74843 Mon Sep 17 00:00:00 2001
From: Sauhard Pande <sauhardp@codeaurora.org>
Date: Thu, 11 May 2017 20:05:50 +0530
Subject: [PATCH 1/3] Camera: CameraHardwareInterface changes to support
 Extended FD

Added QDataCallback definition to pass the extended QCameraFrameMetadata
fields from HIDL to framework.

Change-Id: Idece64f22dd6ee3713c7db658b8583854086a1d8
---
 camera/cameraserver/Android.mk                             |  3 ++-
 services/camera/libcameraservice/Android.mk                |  1 +
 .../libcameraservice/device1/CameraHardwareInterface.cpp   | 14 ++++++++++++++
 .../libcameraservice/device1/CameraHardwareInterface.h     |  7 ++++++-
 4 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/camera/cameraserver/Android.mk b/camera/cameraserver/Android.mk
index b8c94e6..ffc15a8 100644
--- a/camera/cameraserver/Android.mk
+++ b/camera/cameraserver/Android.mk
@@ -30,7 +30,8 @@ LOCAL_SHARED_LIBRARIES := \
 	android.hardware.camera.common@1.0 \
 	android.hardware.camera.provider@2.4 \
 	android.hardware.camera.device@1.0 \
-	android.hardware.camera.device@3.2
+	android.hardware.camera.device@3.2 \
+	vendor.qti.hardware.camera.device@1.0
 
 LOCAL_MODULE:= cameraserver
 LOCAL_32_BIT_ONLY := true
diff --git a/services/camera/libcameraservice/Android.mk b/services/camera/libcameraservice/Android.mk
index 7ec3ccb..410dd33 100644
--- a/services/camera/libcameraservice/Android.mk
+++ b/services/camera/libcameraservice/Android.mk
@@ -77,6 +77,7 @@ LOCAL_SHARED_LIBRARIES:= \
     android.hardware.camera.common@1.0 \
     android.hardware.camera.provider@2.4 \
     android.hardware.camera.device@1.0 \
+    vendor.qti.hardware.camera.device@1.0 \
     android.hardware.camera.device@3.2 \
     android.hardware.camera.device@3.3
 
diff --git a/services/camera/libcameraservice/device1/CameraHardwareInterface.cpp b/services/camera/libcameraservice/device1/CameraHardwareInterface.cpp
index 522d521..28a7a1a 100644
--- a/services/camera/libcameraservice/device1/CameraHardwareInterface.cpp
+++ b/services/camera/libcameraservice/device1/CameraHardwareInterface.cpp
@@ -136,6 +136,20 @@ hardware::Return<void> CameraHardwareInterface::dataCallback(
     return hardware::Void();
 }
 
+hardware::Return<void> CameraHardwareInterface::QDataCallback(
+        DataCallbackMsg msgType, uint32_t data, uint32_t bufferIndex,
+        const vendor::qti::hardware::camera::device::V1_0::QCameraFrameMetadata& metadata) {
+    if (mHidlMemPoolMap.count(data) == 0) {
+        ALOGE("%s: memory pool ID %d not found", __FUNCTION__, data);
+        return hardware::Void();
+    }
+    camera_frame_metadata_t md;
+    md.number_of_faces = metadata.faces.size();
+    md.faces = (camera_face_t*) metadata.faces.data();
+    sDataCb((int32_t) msgType, mHidlMemPoolMap.at(data), bufferIndex, &md, this);
+    return hardware::Void();
+}
+
 hardware::Return<void> CameraHardwareInterface::dataCallbackTimestamp(
         DataCallbackMsg msgType, uint32_t data,
         uint32_t bufferIndex, int64_t timestamp) {
diff --git a/services/camera/libcameraservice/device1/CameraHardwareInterface.h b/services/camera/libcameraservice/device1/CameraHardwareInterface.h
index e519b04..1d58e9b 100644
--- a/services/camera/libcameraservice/device1/CameraHardwareInterface.h
+++ b/services/camera/libcameraservice/device1/CameraHardwareInterface.h
@@ -29,6 +29,7 @@
 #include <hardware/camera.h>
 
 #include <common/CameraProviderManager.h>
+#include <vendor/qti/hardware/camera/device/1.0/IQCameraDeviceCallback.h>
 
 namespace android {
 
@@ -85,7 +86,7 @@ typedef void (*data_callback_timestamp_batch)(
 
 class CameraHardwareInterface :
         public virtual RefBase,
-        public virtual hardware::camera::device::V1_0::ICameraDeviceCallback,
+        public virtual vendor::qti::hardware::camera::device::V1_0::IQCameraDeviceCallback,
         public virtual hardware::camera::device::V1_0::ICameraDevicePreviewCallback {
 
 public:
@@ -395,6 +396,10 @@ private:
             hardware::camera::device::V1_0::DataCallbackMsg msgType,
             const hardware::hidl_vec<
                     hardware::camera::device::V1_0::HandleTimestampMessage>&) override;
+    hardware::Return<void> QDataCallback(
+            hardware::camera::device::V1_0::DataCallbackMsg msgType,
+            uint32_t data, uint32_t bufferIndex,
+            const vendor::qti::hardware::camera::device::V1_0::QCameraFrameMetadata& metadata) override;
 
     /**
      * Implementation of android::hardware::camera::device::V1_0::ICameraDevicePreviewCallback
-- 
2.7.4

